### ISSUE TITLE
Support lambda syntax

### ISSUE THREAD

**Issue by davideicardi** [davideicardi/DynamicExpresso#25]
Support for lambda inside expression, especially to use linq methods like where.

See also: http://stackoverflow.com/questions/17540548/dynamic-expresso-foreach-with-lambda/19010667?noredirect=1#comment44094147_19010667

**Comment by alan-yeung** [comment]
Is this still outstanding? Can I write an expression like this? 

`"myList.Any(x => x == "abc"))"`

**Comment by alan-yeung** [comment]
Thanks for the quick response, David. I'll probably use your suggested workaround in the meantime. Your library is very helpful as I'm using it to interpret expressions stored in the database used by a rule engine.

**Comment by paillave** [comment]
Hey David. It seems you are in lack of motivation for the implementation of this feature. By correcting a bug (recognition of extension methods ignores the caseinsensitive setup), I wanted to look how hard it could be to implement the lambda function feature that I need so badly... It didn't seem too much straight forward to me. Could you mention me in couple of words which you would take to amend your code to achieve at least a basic scenario?

**Comment by davideicardi** [comment]
Hi @paillave . It is not lack of motivation, it is more lack of time ;-) .

Assuming that we want to support `"myList.Any(x => x == "abc"))"` (more complex expressions can be very difficult).

I think that the workflow should be something like:
- recognise that there is a lambda (probably detecting `=>`)
- get the function signature, in the case above the signature of the `Any` function, and extract which parameter we expect. In the case above we should expect something like `Func<string, bool>`.
- In this way we can assume that `x` will be of type `string` and we should returns a `bool`.
- call again the `Interpreter` (in a recursive way) by declaring parameter `x`, get back the delegate and use it for the `Any` function

It is probably a little tricky but it should be possible.

An alternative, as I say often, is to use the official Roslyn scripting from MS (https://github.com/dotnet/roslyn/wiki/Scripting-API-Samples). Probably for complex solution is far better.

**Comment by paillave** [comment]
Thanks for your your answer @davideicardi . I also work on an open source project and I know what it means not to have too much time for it!
About what you mentioned, what puzzles me is that the parser reader always goes forward, then when it realizes it is on an arrow, it realizes too late that the current words wasn't actually a variable but the parameter of the lambda function. I believe I am missing something very obvious, but I can't figure it out. How shall I deal with it?

**Comment by davideicardi** [comment]
Yes, it is a bit tricky. I think that you should modify:

- `NextToken` (see `case '=':`)
- `ParseArgumentList`/`ParseArguments` (this should probably receive the lambda definition as an optional input)
- Inside the `ParseArguments` invoke again a new parser by using the lambda definition as described in the prev post

**Comment by amiqat** [comment]
Any update ? 
this vary useful

**Comment by avidenic** [comment]
Hi, 

has anyone had any time to work on this? It has been 5 years =)

Best regards

**Comment by amoorthaemer** [comment]
@metoule

Hi, I noticed that you actively working on a forked branch of Dynamic Expresso. I am looking into using this package, but the fact that it doesn't supports Linq methods that accept a lamba function makes it a no-go for me. Are you currently or planning to look into this missing feature? Please let me know
TIA Bert

**Comment by metoule** [comment]
@amoorthaemer

Hi Bert, 
for now, I'm only trying to add the features I need for my own project, and I don't need to support lambda expressions.

That being said, I find this project quite enjoyable to work on, so I might as well have a look at this issue. Please understand that I make no promise, but who knows :)

**Comment by metoule** [comment]
@amoorthaemer 

Hi Bert (and all the watchers), 
I have a working prototype! It still needs some work, but I was able to successfully run the following code:

````C#
[Test]
public void Select_inferred_return_type()
{
  var target = new Interpreter();
  var list = new List<int> { 1, 10, 19, 21 };
  target.SetVariable("myList", list);
  
  var results = target.Eval<IEnumerable<char>>("myList.Select(i => i.ToString()).Select(str => str[0])");
  
  Assert.AreEqual(4, results.Count());
  Assert.AreEqual(new[] { '1', '1', '1', '2' }, results);
}
````

Any example of the kind of lambda you'd want to execute would be helpful. 

Note the following limitations:
- the lambda body can't be a block
- it can't return a new object (no `.Select(i => new { Field = i })`)

**Comment by amoorthaemer** [comment]
@metoule 

Nice! Well my requirements are simple (for now ;) ) but it would be nice to be able to access parent scope parameters, so complex LINQ queries can be executed also.

If you need some testing I'm happy to help out

/Cheers

Bert

**Comment by metoule** [comment]
Yes, you can return `new SomeType()`, but you can't use the object initializer syntax to set its properties: every property you want to set will have to be passed in the constructor.

So for example:

```csharp
class SomeType
{
	public string MyStr { get; set; }
	public SomeType(string str)
	{
		MyStr = str;
	}
}

[Test]
public void Select_new()
{
	var list = new List<string> { "this", "is", "awesome" };
	var target = new Interpreter()
				.EnableLambdaExpressions(true)
				.Reference(typeof(SomeType))
				.SetVariable("myList", list);

	var results = target.Eval<IEnumerable<SomeType>>("myList.Select(str => new SomeType(str))");
}
```

but not 

```csharp
	var results = target.Eval<IEnumerable<SomeType>>("myList.Select(str => new SomeType { MyStr = str })");
```

Note that this isn't directly related to lambda expressions support. Supporting the object initializer syntax is another topic :)

**Comment by metoule** [comment]
I've created PR #141 to support the object initializer syntax.
